### PR TITLE
mgr/rook: Blinking lights

### PR DESCRIPTION
**Pull request by jmolmo**
Blinking lights implementation

In my "dev" environment ( only vms) I have tested the new `rook_cluster.blink_light` method calling it manually with several different "locs" parameter.
But I need to have baremetal machines where this can be tested properly and switching on/off the disks identification lights can be verified.

My aim now is to make a integration test of this functionality to run in the Rook CI. But in parallel i will try to locate and to verify this on a real cluster.

Signed-off-by: Juan Miguel Olmo Martínez <jolmomar@redhat.com>

### CODE REVIEW

File: src/pybind/mgr/rook/module.py [hunk 1]
```diff
@@ -133,13 +139,15 @@ def __init__(self, *args, **kwargs):
...
-        self._k8s = None
+        self._k8s_CoreV1_api = None
+        self._k8s_BatchV1_api = None
         self._rook_cluster = None
         self._rook_env = RookEnv()
 
         self._shutdown = threading.Event()
 
-        self.all_progress_references = list()  # type: List[orchestrator.ProgressReference]
+        # type: List[orchestrator.ProgressReference]
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py]:
```
mypy run-test-pre: PYTHONHASHSEED='1634021868'
mypy run-test: commands[0] | mypy --config-file=../../mypy.ini cephadm/module.py mgr_module.py mgr_util.py orchestrator.py orchestrator_cli/module.py progress/module.py rook/module.py test_orchestrator/module.py
rook/module.py:149: error: misplaced type annotation
Found 1 error in 1 file (checked 8 source files)
```

File: src/pybind/mgr/rook/module.py [hunk 2]
```diff
@@ -55,7 +56,8 @@ def wrapper(*args, **kwargs):
 def write_completion(on_complete,  # type: Callable
                      message,  # type: str
                      mgr,
-                     calc_percent=None  # type: Optional[Callable[[], RookCompletion]]
+                     # type: Optional[Callable[[], RookCompletion]]
+                     calc_percent=None
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py]: Hm. That probably prevents the type annotation from being checked. 

```suggestion
                     calc_percent: Optional[Callable[[], RookCompletion]]=None
```

File: src/pybind/mgr/rook/module.py [hunk 3]
```diff
@@ -72,10 +74,13 @@ def __init__(self):
         self.namespace = os.environ.get('POD_NAMESPACE', 'rook-ceph')
 
         # ROOK_CEPH_CLUSTER_CRD_NAME is new is Rook 1.0
-        self.cluster_name = os.environ.get('ROOK_CEPH_CLUSTER_CRD_NAME', self.namespace)
+        self.cluster_name = os.environ.get(
+            'ROOK_CEPH_CLUSTER_CRD_NAME', self.namespace)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py]: could you move those style changes into a different commit?

File: src/pybind/mgr/rook/rook_cluster.py [hunk 4]
```diff
@@ -180,21 +179,22 @@ def _watch(self, res_ver):
 
 
 class RookCluster(object):
-    def __init__(self, k8s, rook_env):
+    def __init__(self, coreV1_api, batchV1_api, rook_env):
         self.rook_env = rook_env  # type: RookEnv
-        self.k8s = k8s  # type: CoreV1Api
+        self.coreV1_api = coreV1_api
+        self.batchV1_api = batchV1_api
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/rook_cluster.py]:
```suggestion
        self.coreV1_api: CoreV1Api=coreV1_api
        self.batchV1_api: BatchV1Api=batchV1_api
```
This really helps when using an IDE

File: src/pybind/mgr/rook/rook_cluster.py [hunk 5]
```diff
@@ -559,3 +559,89 @@ def _patch(self, crd, crd_name, cr_name, func):
...
+        image = "ceph/ceph:v14.2.6"
+
+        try:
+            api_response = self.coreV1_api.list_namespaced_pod(self.rook_env.namespace, label_selector="app=rook-ceph-mon", timeout_seconds=10)
+            if api_response.items:
+                image = api_response.items[-1].spec.containers[0].image
+        except ApiException as e:
+            log.warning("Error getting ceph image: {}, using <{}> as default".format(e, image))
+
+        return image
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/rook_cluster.py]: another idea would be to fetch this from the CephCluster CR?

File: src/pybind/mgr/rook/rook_cluster.py [hunk 6]
```diff
@@ -559,3 +559,89 @@ def _patch(self, crd, crd_name, cr_name, func):
...
+
+    def blink_light(self, ident_fault, on, locs):
+        # type: (str, bool, List[DeviceLightLoc]) -> string
+        operation_id = str(hash(locs))
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/rook_cluster.py]: should we add `ident_fault` and `on` to the hash?

File: src/pybind/mgr/rook/rook_cluster.py [hunk 7]
```diff
@@ -559,3 +559,89 @@ def _patch(self, crd, crd_name, cr_name, func):
...
+                api_response = self.batchV1_api.delete_namespaced_job(operation_id, self.rook_env.namespace, propagation_policy="Background")
+            except ApiException as e:
+                if e.status != 404: # No problem if the job does not exist
+                    raise
+
+            # wait until the job is not present
+            deleted = False
+            while not deleted:
+                api_response = self.batchV1_api.list_namespaced_job(self.rook_env.namespace, label_selector="ident=%s" % operation_id, timeout_seconds=10)
+                deleted = not api_response.items
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/rook_cluster.py]: do we need a timeout?

File: src/pybind/mgr/rook/module.py (hunk 1 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/module.py]: Merge mischief ... leaved as it was.

File: src/pybind/mgr/rook/module.py (hunk 2 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/module.py]: withe the suggested change i hava "invalid syntax ..." .. i leave as it was

File: src/pybind/mgr/rook/module.py (hunk 3 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/module.py]: Yes. Sorry ... my IDE doing things by itself ... i will revert all these changes

File: src/pybind/mgr/rook/rook_cluster.py (hunk 4 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: Sorry... PEP 484 ..... i was going to say something not correct about this "masked beta feature"  but after 50 years i have achieve to have a good self-control.
I only will say. I don't like to do that.

File: src/pybind/mgr/rook/rook_cluster.py (hunk 5 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: Why is better to get it from the cluster CR?

File: src/pybind/mgr/rook/rook_cluster.py (hunk 6 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: I think this will keep safe the last "light" operation executed over a physical disk.  In the case of several operations launched simultaneously or almost simultaneously, probably the right thing to do is to deem as valid only the last one. (independently if this operation is to switch on or off the light). 
And as we use only the identification light  neither is needed to add the type of operation to the hash

File: src/pybind/mgr/rook/rook_cluster.py (hunk 7 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: yes... better be nice with the CPU.

File: src/pybind/mgr/rook/rook_cluster.py (hunk 4 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py]: we're finally on Python 3 :tada: :tada: :tada: :tada:

File: src/pybind/mgr/rook/rook_cluster.py (hunk 5 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py]: In case of an upgrade scenario, the pod image might be the old image. The CephCluster image should be the "correct" one.

File: src/pybind/mgr/rook/rook_cluster.py (hunk 7 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py]: Hm. What about if, of some unknown reason, the resource is never getting deleted?

File: src/pybind/mgr/rook/rook_cluster.py (hunk 5 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: if it is the old image, there are lot of possibilities that the OSDs will have also the same old image, and what i want to do is to use a Ceph image already present in the node.
If the mon has an updated image, and the OSDs an older image, this should be updated, turn off/on a light will be even positive because we are going to download the new image before the upgrade procedure, doing this more fast.

I think that to use the mon image is the best  choice.

Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py]: ok with me. But then, can we please fail, instead of using `"ceph/ceph:v14.2.6"` by default?

File: src/pybind/mgr/rook/rook_cluster.py (hunk 7 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: if we reach this line.. the delete of the job has been launched without any error, so k8s is just doing the task. If the k8s cannot finish with the task then what we have is a bug in k8s .... but apart of that. We have a second safety mechanism, that is that the job execution is limited to 60 seconds, and after finished (fail or succes) the job must be automatically deleted in 10 seconds. 
Again If k8s cannot delete the job we will have a bug in k8s ...

If we abort without deleting the job. (because the k8s bug), we are going to move the problem to other place , because if a job for a disk cannot be deleted .. then the creation of a new job for that disk is not going to be possible, and then we will have another different error caused because the problem of deletion.

Maybe what we can do is to add a retry-limit checking if the job has been deleted and abort if we reach this limit.. what do you think?

Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py]: sounds good! The problem with endless loops is, the orch cli hangs indefinitely making it much harder to understand the underlying problem.

Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: Done!

File: src/pybind/mgr/rook/rook_cluster.py (hunk 5 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: Ok. Done.

File: src/pybind/mgr/rook/module.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/module.py]: ping?

**Comment by sebastian-philipp** [comment]
```
mypy run-test-pre: PYTHONHASHSEED='685483167'
mypy run-test: commands[0] | mypy --config-file=../../mypy.ini cephadm/module.py mgr_module.py mgr_util.py orchestrator/__init__.py progress/module.py rook/module.py test_orchestrator/module.py
rook/module.py:34: note: In module imported here:
rook/rook_cluster.py: note: In member "_fetch" of class "KubernetesResource":
rook/rook_cluster.py:110: error: Name 'V1ListMeta' is not defined
rook/rook_cluster.py: note: In function "describe_pods":
rook/rook_cluster.py:288: error: Name 'V1Pod' is not defined
rook/rook_cluster.py: note: In member "blink_light" of class "RookCluster":
rook/rook_cluster.py:578: error: Name 'DeviceLightLoc' is not defined
rook/rook_cluster.py:578: error: Name 'string' is not defined
rook/rook_cluster.py:592: error: "List[Any]" has no attribute "path"
rook/rook_cluster.py:592: error: "List[Any]" has no attribute "dev"
rook/rook_cluster.py:596: error: "List[Any]" has no attribute "host"
rook/rook_cluster.py:632: error: "List[Any]" has no attribute "host"
rook/rook_cluster.py:632: error: "List[Any]" has no attribute "path"
rook/rook_cluster.py:632: error: "List[Any]" has no attribute "dev"
rook/module.py: note: In member "blink_device_light" of class "RookOrchestrator":
rook/module.py:452: error: Name 'DeviceLightLoc' is not defined
rook/module.py:458: error: "List[Any]" has no attribute "dev"
Found 12 errors in 2 files (checked 7 source files)
```

File: src/pybind/mgr/rook/module.py [hunk 8]
```diff
@@ -449,3 +448,13 @@ def has_osds(all_hosts):
 
         c = self.get_hosts().then(execute)
         return c
+
+    def blink_device_light(self, ident_fault, on, locs):
+        # type: (str, bool, List[DeviceLightLoc]) -> RookCompletion
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py]: `rook/module.py:452: error: Name 'DeviceLightLoc' is not defined`

File: src/pybind/mgr/rook/rook_cluster.py [hunk 9]
```diff
@@ -27,9 +28,8 @@
     pass  # just for type annotations
 
 try:
+    from kubernetes import client, watch
     from kubernetes.client.rest import ApiException
-    from kubernetes.client import V1ListMeta, CoreV1Api, V1Pod
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/rook_cluster.py]: this change makes mypy fail.

File: src/pybind/mgr/rook/rook_cluster.py [hunk 10]
```diff
@@ -560,3 +561,110 @@ def _patch(self, crd, crd_name, cr_name, func):
...
+            if api_response.items:
+                return api_response.items[-1].spec.containers[0].image
+            else:
+                raise orchestrator.OrchestratorError(
+                        "Error getting ceph image. Cluster without monitors")
+        except ApiException as e:
+            log.error("Error getting ceph image: {}".format(e))
+
+    def blink_light(self, ident_fault, on, locs):
+        # type: (str, bool, List[DeviceLightLoc]) -> string
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/rook_cluster.py]: `rook/rook_cluster.py:578: error: Name 'string' is not defined`

File: src/pybind/mgr/rook/rook_cluster.py [hunk 11]
```diff
@@ -560,3 +561,110 @@ def _patch(self, crd, crd_name, cr_name, func):
...
+        # job definition
+        job_metadata = client.V1ObjectMeta(name=operation_id,
+                                           namespace= self.rook_env.namespace,
+                                           labels={"ident": operation_id})
+        pod_metadata = client.V1ObjectMeta(labels={"ident": operation_id})
+        pod_container = client.V1Container(name="ceph-lsmcli-command",
+                                           image=self.get_ceph_image(),
+                                           command=["lsmcli",],
+                                           args=['local-disk-%s-led-%s' % (ident_fault,'on' if on else 'off'),
+                                                 '--path', locs.path or locs.dev,])
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/rook_cluster.py]: `rook/rook_cluster.py:592: error: "List[Any]" has no attribute "path"`

**Comment by sebastian-philipp** [comment]
you can run mypy locally as well:

```
cd /src/pybind/mgr ; ../../script/run_tox.sh --tox-envs=py3,mypy mgr
```

File: src/pybind/mgr/rook/rook_cluster.py (hunk 9 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: it works for me ... 
[jolmomar@juanmipc mgr]$ mypy --config-file=../../mypy.ini  rook/module.py  rook/rook_cluster.py
Success: no issues found in 2 source files

**Comment by jmolmo** [comment]
> you can run mypy locally as well:
> 
> ```
> cd /src/pybind/mgr ; ../../script/run_tox.sh --tox-envs=py3,mypy mgr
> ```

Thx!. 
I have seen that i can also execute mypy directly over the required source files.

File: src/pybind/mgr/rook/rook_cluster.py [hunk 12]
```diff
@@ -107,7 +107,7 @@ def _fetch(self):
         """ Execute the requested api method as a one-off fetch"""
         response = self.api_func(**self.kwargs)
         # metadata is a V1ListMeta object type
-        metadata = response.metadata  # type: V1ListMeta
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/rook_cluster.py]: Please revert this!

File: src/pybind/mgr/rook/rook_cluster.py [hunk 13]
```diff
@@ -285,7 +286,6 @@ def describe_pods(self, service_type, service_id, nodename):
                         rook_file_system=<self.fs_name>
         """
         def predicate(item):
-            # type: (V1Pod) -> bool
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/rook_cluster.py]: Please revert this!

**Comment by jmolmo** [comment]
I think that not all the types can be annotated properly.. or i do not know how to do that....if not is possible to reference  "alien" types with mypy ... is a pity, because this implies to lose interesting comments about  types that we had before "mypy" and the awesome PEP  484 has been applied as a new code purity check. :-)

**Comment by sebastian-philipp** [comment]
> I think that not all the types can be annotated properly.. 

Sure, but I'm a bit :-1:  for artificially reducing the tesing coverage. Those two type annotations are checked by mypy already! Removing them seems a bit odd to me.

**Comment by jmolmo** [comment]
> > I think that not all the types can be annotated properly..
> 
> Sure, but I'm a bit  for artificially reducing the tesing coverage. Those two type annotations are checked by mypy already! Removing them seems a bit odd to me.

If i do not remove these two type annotations I have:

```
[jolmomar@juanmipc mgr]$ mypy --config-file=../../mypy.ini  rook/module.py  rook/rook_cluster.py
rook/rook_cluster.py: note: In member "_fetch" of class "KubernetesResource":
rook/rook_cluster.py:110: error: Name 'V1ListMeta' is not defined
rook/rook_cluster.py: note: In function "describe_pods":
rook/rook_cluster.py:288: error: Name 'V1Pod' is not defined
Found 2 errors in 1 file (checked 2 source files)
```
Do you know how to tell _mypy_ about the nature of V1Pod and V1ListMeta?

**Comment by sebastian-philipp** [comment]
> > > I think that not all the types can be annotated properly..
> > 
> > 
> > Sure, but I'm a bit  for artificially reducing the tesing coverage. Those two type annotations are checked by mypy already! Removing them seems a bit odd to me.
> 
> If i do not remove these two type annotations I have:
> 
> ```
> [jolmomar@juanmipc mgr]$ mypy --config-file=../../mypy.ini  rook/module.py  rook/rook_cluster.py
> rook/rook_cluster.py: note: In member "_fetch" of class "KubernetesResource":
> rook/rook_cluster.py:110: error: Name 'V1ListMeta' is not defined
> rook/rook_cluster.py: note: In function "describe_pods":
> rook/rook_cluster.py:288: error: Name 'V1Pod' is not defined
> Found 2 errors in 1 file (checked 2 source files)
> ```
> 
> Do you know how to tell _mypy_ about the nature of V1Pod and V1ListMeta?

Yes! This is the import statement you removed! https://github.com/ceph/ceph/pull/33366/files#diff-45db35bc78bc3027f9421bd371b19e9bL31

**Comment by jmolmo** [comment]
> > > > I think that not all the types can be annotated properly..
> > > 
> > > 
> > > Sure, but I'm a bit  for artificially reducing the tesing coverage. Those two type annotations are checked by mypy already! Removing them seems a bit odd to me.
> > 
> > 
> > If i do not remove these two type annotations I have:
> > ```
> > [jolmomar@juanmipc mgr]$ mypy --config-file=../../mypy.ini  rook/module.py  rook/rook_cluster.py
> > rook/rook_cluster.py: note: In member "_fetch" of class "KubernetesResource":
> > rook/rook_cluster.py:110: error: Name 'V1ListMeta' is not defined
> > rook/rook_cluster.py: note: In function "describe_pods":
> > rook/rook_cluster.py:288: error: Name 'V1Pod' is not defined
> > Found 2 errors in 1 file (checked 2 source files)
> > ```
> > 
> > 
> > Do you know how to tell _mypy_ about the nature of V1Pod and V1ListMeta?
> 
> Yes! This is the import statement you removed! https://github.com/ceph/ceph/pull/33366/files#diff-45db35bc78bc3027f9421bd371b19e9bL31

ahhhh!!!. Sorry ... fixed!

File: src/pybind/mgr/rook/rook_cluster.py (hunk 11 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py]: yeah, removing the type annotation silenced the error message, but `locs` is still of type `List[DeviceLightLoc]`, which still doesn't have a `path` attribute! The type annotations were correct, just the code is wrong.

File: src/pybind/mgr/rook/module.py [hunk 14]
```diff
@@ -103,7 +104,8 @@ def process(self, completions):
         # type: (List[RookCompletion]) -> None
 
         if completions:
-            self.log.info("process: completions={0}".format(orchestrator.pretty_print(completions)))
+            self.log.info("process: completions={0}".format(
+                orchestrator.pretty_print(completions)))
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py]: would you mind, moving those changed into a different commit?

File: src/pybind/mgr/rook/module.py [hunk 15]
```diff
@@ -291,29 +296,26 @@ def list_daemons(self, daemon_type=None, daemon_id=None, node_name=None, refresh
...
         # type: (orchestrator.NFSServiceSpec) -> RookCompletion
-        return self._service_add_decorate("NFS", spec,
-                                          self.rook_cluster.add_nfsgw)
+        return self._service_add_decorate("NFS", spec, self.rook_cluster.add_nfsgw)
 
     def _service_rm_decorate(self, typename, name, func):
         return write_completion(
-            on_complete=lambda : func(name),
+            on_complete=lambda: func(name),
             message="Removing {} services for {}".format(typename, name),
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py]: also into a different commit?

File: src/pybind/mgr/rook/module.py [hunk 16]
```diff
@@ -325,12 +327,10 @@ def remove_service(self, service_type, service_name):
...
-            return self._service_rm_decorate(
-                'RGW', service_name, lambda: self.rook_cluster.rm_service('cephobjectstores', service_name)
-            )
+            return self._service_rm_decorate('RGW', service_name, lambda: self.rook_cluster.rm_service('cephobjectstores', service_name))
+
         elif service_type == 'nfs':
-            return self._service_rm_decorate(
-                'NFS', service_name, lambda: self.rook_cluster.rm_service('cephnfses', service_name)
+            return self._service_rm_decorate('NFS', service_name, lambda: self.rook_cluster.rm_service('cephnfses', service_name)
             )
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py]: also into this "style fixes" commit?

File: src/pybind/mgr/rook/module.py [hunk 17]
```diff
@@ -405,19 +405,18 @@ def execute(all_hosts_):
...
             pod_osd_ids = set()
-            pods = self.k8s.list_namespaced_pod(self._rook_env.namespace,
-                                                 label_selector="rook_cluster={},app=rook-ceph-osd".format(self._rook_env.cluster_name),
-                                                 field_selector="spec.nodeName={0}".format(
-                                                     drive_group.hosts(all_hosts)[0]
-                                                 )).items
+            pods = self.k8s.list_namespaced_pod(self._rook_env.namespace, label_selector="rook_cluster={},app=rook-ceph-osd".format(self._rook_env.cluster_name),
+                                                field_selector="spec.nodeName={0}".format(
+                                                    drive_group.hosts(all_hosts)[0]
+                                                )).items
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py]: also into this style-fixes commit?

File: src/pybind/mgr/rook/rook_cluster.py (hunk 9 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py]: this import is still missing

Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: This import has been replaced by:
`from kubernetes import client, watch`
Do you prefer to have imported explicitly CoreV1Api, V1Pod? ( V1ListMeta only appears in comments)

**Comment by jmolmo** [comment]
The import has been replaced by "from kubernetes import client"

**Comment by sebastian-philipp** [comment]
I'd like to push to your repo. could you enable this flag? https://github.blog/2016-09-07-improving-collaboration-with-forks/

**Comment by jmolmo** [comment]
> I'd like to push to your repo. could you enable this flag? https://github.blog/2016-09-07-improving-collaboration-with-forks/

It was already enabled.

**Comment by sebastian-philipp** [comment]
ping? could you please apply

```diff
From ff90c8335658c6668bf858dc1a555b4a56ab9d21 Mon Sep 17 00:00:00 2001
From: Sebastian Wagner <sebastian.wagner@suse.com>
Date: Fri, 21 Feb 2020 12:41:42 +0100
Subject: [PATCH] mgr/rook: Add type annotations for blink_light

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>
---
 src/pybind/mgr/rook/module.py       | 1 +
 src/pybind/mgr/rook/rook_cluster.py | 2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/pybind/mgr/rook/module.py b/src/pybind/mgr/rook/module.py
index 098ef66280..8364873ad2 100644
--- a/src/pybind/mgr/rook/module.py
+++ b/src/pybind/mgr/rook/module.py
@@ -447,6 +447,7 @@ class RookOrchestrator(MgrModule, orchestrator.Orchestrator):
         return c
 
     def blink_device_light(self, ident_fault, on, locs):
+        # type: (str, bool, List[orchestrator.DeviceLightLoc]) -> RookCompletion
         return write_completion(
             on_complete=lambda: self.rook_cluster.blink_light(
                 ident_fault, on, locs),
diff --git a/src/pybind/mgr/rook/rook_cluster.py b/src/pybind/mgr/rook/rook_cluster.py
index c08c6aefd7..06ffabcbc6 100644
--- a/src/pybind/mgr/rook/rook_cluster.py
+++ b/src/pybind/mgr/rook/rook_cluster.py
@@ -575,7 +575,7 @@ class RookCluster(object):
         except ApiException as e:
             log.error("Error getting ceph image: {}".format(e))
 
-    def blink_light(self, ident_fault, on, locs):
+    def blink_light(self, ident_fault: str, on: bool, locs: List[orchestrator.DeviceLightLoc]):
         operation_id = str(hash(locs))
         message = ""
 
-- 
2.17.1
```

**Comment by jmolmo** [comment]
We are using in this PR Python 3 type annotations in the functions, but all the others type annotations in other functions are Python 2.
I think that probably we should change all of them to Python3... in a different PR only for that. What do ypou think @sebastian-philipp ?

File: src/pybind/mgr/rook/rook_cluster.py [hunk 18]
```diff
@@ -560,3 +561,113 @@ def _patch(self, crd, crd_name, cr_name, func):
...
+            api_response = self.batchV1_api.delete_namespaced_job(operation_id,
+                                                                  self.rook_env.namespace,
+                                                                  propagation_policy="Background")
+        except ApiException as e:
+            if e.status != 404: # No problem if the job does not exist
+                raise
+
+        return message
+
+    def blink_light(self, ident_fault: str, on: bool, locs: List[orchestrator.DeviceLightLoc]) -> List[str]:
```
Comment (tchaikov) [inline src/pybind/mgr/rook/rook_cluster.py]:
```
2020-02-29T11:45:43.588 INFO:tasks.ceph.mgr.y.smithi061.stderr:2020-02-29T11:45:43.585+0000 7f5cb714aec0 -1 mgr[py] Module not found: 'rook'
2020-02-29T11:45:43.589 INFO:tasks.ceph.mgr.y.smithi061.stderr:2020-02-29T11:45:43.585+0000 7f5cb714aec0 -1 mgr[py] Traceback (most recent call last):
2020-02-29T11:45:43.589 INFO:tasks.ceph.mgr.y.smithi061.stderr:  File "/usr/share/ceph/mgr/rook/__init__.py", line 2, in <module>
2020-02-29T11:45:43.589 INFO:tasks.ceph.mgr.y.smithi061.stderr:    from .module import RookOrchestrator
2020-02-29T11:45:43.589 INFO:tasks.ceph.mgr.y.smithi061.stderr:  File "/usr/share/ceph/mgr/rook/module.py", line 33, in <module>
2020-02-29T11:45:43.590 INFO:tasks.ceph.mgr.y.smithi061.stderr:    from .rook_cluster import RookCluster
2020-02-29T11:45:43.590 INFO:tasks.ceph.mgr.y.smithi061.stderr:  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 183, in <module>
2020-02-29T11:45:43.590 INFO:tasks.ceph.mgr.y.smithi061.stderr:    class RookCluster(object):
2020-02-29T11:45:43.590 INFO:tasks.ceph.mgr.y.smithi061.stderr:  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 672, in RookCluster
2020-02-29T11:45:43.591 INFO:tasks.ceph.mgr.y.smithi061.stderr:    def blink_light(self, ident_fault: str, on: bool, locs: List[orchestrator.DeviceLightLoc]) -> List[str]:
2020-02-29T11:45:43.591 INFO:tasks.ceph.mgr.y.smithi061.stderr:NameError: name 'List' is not defined
2020-02-29T11:45:43.591 INFO:tasks.ceph.mgr.y.smithi061.stderr:
```

http://pulpito.ceph.com/kchai-2020-02-29_11:07:48-rados-wip-kefu-testing-2020-02-29-1522-distro-basic-smithi/4813422/

**Comment by tchaikov** [comment]
@jmolmo ping?

File: src/pybind/mgr/rook/rook_cluster.py (hunk 18 above)
Comment (jmolmo) [reply src/pybind/mgr/rook/rook_cluster.py]: Yeah. Thanks. 
It seems that the mypy embedded notation is not ok. 
Replaced by comments

**Comment by jmolmo** [comment]
Tested in a real baremetal cluster:

Execution example:

```

[root@rook-ceph-tools-7f96779fb9-xkpd4 /]# ceph device ls
DEVICE                               HOST:DEV   DAEMONS  LIFE EXPECTANCY
WDC_WDS200T2B0A-00SM50_183503800976  cpach:sdo  osd.0 

[root@rook-ceph-tools-7f96779fb9-xkpd4 /]# ceph device light on WDC_WDS200T2B0A-00SM50_183503800976
['']
[root@rook-ceph-tools-7f96779fb9-xkpd4 /]# ceph device light off WDC_WDS200T2B0A-00SM50_183503800976
['']
```


In case of errors:
```

[root@rook-ceph-tools-7f96779fb9-xkpd4 /]# ceph device light on WDC_WDS200T2B0A-00SM50_183503800976
_exception: (422)
Reason: Unprocessable Entity
HTTP response headers: HTTPHeaderDict({'Content-Type': 'application/json', 'Date': 'Fri, 13 Mar 2020 16:36:44 GMT', 'Content-Length': '447'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Job.batch \"3822060501990855105\" is invalid: spec.template.spec.containers[0].volumeMounts[1].name: Not found: \"runudev\"","reason":"Invalid","details":{"name":"3822060501990855105","group":"batch","kind":"Job","causes":[{"reason":"FieldValueNotFound","message":"Not found: \"runudev\"","field":"spec.template.spec.containers[0].volumeMounts[1].name"}]},"code":422}


[root@rook-ceph-tools-7f96779fb9-xkpd4 /]# ceph device light on WDC_WDS200T2B0A-00SM50_183503800976
['LIB_BUG(1): BUG: Failed to open /dev/disk/by-path/pci-0000:82:00.0-sas-exp0x50030480091072bf-phy31-lun-0, error: 1, Operation not permitted \n']
```

File: src/pybind/mgr/rook/rook_cluster.py (hunk 18 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/rook_cluster.py]:
```python
from typing import List
```
might be missing?

**Comment by jmolmo** [comment]
In order to work this modification is needed to add new rights to the ceph manager module in rook:

See: [Ceph: Added rook manager module rights for read pod logs](https://github.com/rook/rook/pull/5025)

**Comment by sebastian-philipp** [comment]
```
rook/module.py:35: note: In module imported here:
rook/rook_cluster.py: note: In member "remove_pods" of class "RookCluster":
rook/rook_cluster.py:362: error: "RookCluster" has no attribute "k8s"
rook/rook_cluster.py:365: error: Name 'V1DeleteOptions' is not defined
```

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2020-03-24_03:31:20-rados-wip-kefu-testing-2020-03-23-2313-distro-basic-smithi/

failures are not related.

- http://pulpito.ceph.com/kchai-2020-03-24_03:31:20-rados-wip-kefu-testing-2020-03-23-2313-distro-basic-smithi/4885988: already fixed in cbt
- http://pulpito.ceph.com/kchai-2020-03-24_03:31:20-rados-wip-kefu-testing-2020-03-23-2313-distro-basic-smithi/4886001/: already fixed in dashboard in ceph
- http://pulpito.ceph.com/kchai-2020-03-24_03:31:20-rados-wip-kefu-testing-2020-03-23-2313-distro-basic-smithi/4886049: tracked by https://tracker.ceph.com/issues/23947

**Comment by tchaikov** [comment]
@jmolmo needs rebase

**Comment by tchaikov** [comment]
@jmolmo could remove the "merge" commit in this PR?
